trajopt/src/collision_terms.cpp

```diff
@@ -159,6 +159,225 @@ void DebugPrintInfo(const tesseract_collision::ContactResult& res,
   std::printf("\n");
 }
 
+GradientResults CollisionEvaluator::GetGradient(const Eigen::VectorXd& dofvals,
+                                                const tesseract_collision::ContactResult& contact_result,
+                                                const Eigen::Vector2d& data,
+                                                bool isTimestep1)
+{
+  GradientResults results(data);
+  tesseract_environment::AdjacencyMapPair::ConstPtr itA = adjacency_map_->getLinkMapping(contact_result.link_names[0]);
+  if (itA != nullptr)
+  {
+    results.gradients[0].has_gradient = true;
+
+    Eigen::MatrixXd jac;
+    jac.resize(6, manip_->numJoints());
+
+    // Calculate Jacobian
+    manip_->calcJacobian(jac, dofvals, itA->link_name);
+
+    // Need to change the base and ref point of the jacobian.
+    // When changing ref point you must provide a vector from the current ref
+    // point to the new ref point.
+    results.gradients[0].scale = 1;
+    Eigen::Isometry3d link_transform = contact_result.transform[0];
+    if (contact_result.cc_type[0] != tesseract_collision::ContinuousCollisionType::CCType_None)
+    {
+      assert(contact_result.cc_time[0] >= 0.0 && contact_result.cc_time[0] <= 1.0);
+      results.gradients[0].scale = (isTimestep1) ? contact_result.cc_time[0] : (1 - contact_result.cc_time[0]);
+      link_transform = (isTimestep1) ? contact_result.cc_transform[0] : contact_result.transform[0];
+    }
+    tesseract_kinematics::jacobianChangeBase(jac, world_to_base_);
+    tesseract_kinematics::jacobianChangeRefPoint(jac,
+                                                 (link_transform * itA->transform.inverse()).linear() *
+                                                     (itA->transform * contact_result.nearest_points_local[0]));
+
+    // Eigen::Isometry3d test_link_transform, temp1, temp2;
+    // manip->calcFwdKin(test_link_transform, dofvals, itA->link_name);
+    // temp1 = world_to_base * test_link_transform;
+    // temp2 = link_transform * itA->transform.inverse();
+    // assert(temp1.isApprox(temp2, 0.0001));
+
+    // Eigen::MatrixXd jac_test;
+    // jac_test.resize(6, manip->numJoints());
+    // tesseract_kinematics::numericalJacobian(jac_test, world_to_base, *manip, dofvals, itA->link_name,
+    // res.nearest_points_local[0]); bool check = jac.isApprox(jac_test, 1e-3); assert(check == true);
```

Comment:
Consider removing commented code if it is no longer useful

Reply:
I use it periodically as a sanity check when I make changes to this function.